packages/flutter/lib/src/widgets/focus_manager.dart

```diff
@@ -1844,6 +1674,219 @@ class FocusManager with DiagnosticableTreeMixin, ChangeNotifier {
   }
 }
 
+// A class to detect and manage the highlight mode transitions.
+// An instance of this is owned by the FocusManager.
+class _HighlightModeManager {
+  _HighlightModeManager();
+
+  // If set, indicates if the last interaction detected was touch or not.
+  // If null, no interactions have occurred yet.
+  bool? _lastInteractionWasTouch;
+
+  FocusHighlightMode get highlightMode => _highlightMode ?? _defaultModeForPlatform;
+  FocusHighlightMode? _highlightMode;
+
+  FocusHighlightStrategy get strategy => _strategy;
+  FocusHighlightStrategy _strategy = FocusHighlightStrategy.automatic;
+  set strategy(FocusHighlightStrategy value) {
+    if (_strategy == value) {
+      return;
+    }
+    _strategy = value;
+    updateMode();
+  }
+
+  /// Register a closure to be called when the [FocusManager] notifies its listeners
+  /// that the value of [highlightMode] has changed.
+  void addListener(ValueChanged<FocusHighlightMode> listener) => _listeners.add(listener);
+
+  /// Remove a previously registered closure from the list of closures that the
+  /// [FocusManager] notifies.
+  void removeListener(ValueChanged<FocusHighlightMode> listener) => _listeners.remove(listener);
+
+  // The list of listeners for [highlightMode] state changes.
+  HashedObserverList<ValueChanged<FocusHighlightMode>> _listeners = _emptyListeners;
+  static final HashedObserverList<ValueChanged<FocusHighlightMode>> _emptyListeners = HashedObserverList<ValueChanged<FocusHighlightMode>>();
+
+  void registerGlobalHandlers() {
+    assert(ServicesBinding.instance.keyEventManager.keyMessageHandler == null);
+    ServicesBinding.instance.keyEventManager.keyMessageHandler = handleKeyMessage;
+    GestureBinding.instance.pointerRouter.addGlobalRoute(handlePointerEvent);
+  }
+
+  @mustCallSuper
+  void dispose() {
+    if (ServicesBinding.instance.keyEventManager.keyMessageHandler == handleKeyMessage) {
+      GestureBinding.instance.pointerRouter.removeGlobalRoute(handlePointerEvent);
+      ServicesBinding.instance.keyEventManager.keyMessageHandler = null;
+    }
+    _listeners = _emptyListeners;
+  }
+
+  @pragma('vm:notify-debugger-on-exception')
+  void notifyListeners() {
```

Comment:
Is there a reason this class can't just extend/mixin ChangeNotifier instead of re-implementing that functionality?

Reply:
If there is a reason, maybe document it?

Reply:
Yes, because ChangeNotifier doesn't provide the modified value in the callback. I'll add some docs.